tasks/notification_orchestrator.py

```diff
@@ -0,0 +1,631 @@
+class NotificationOrchestratorTask(
+    BaseCodecovTask, name=notification_orchestrator_task_name
+):
+    def run_impl(
+        self,
+        db_session: Session,
+        *args,
+        repoid: int,
+        commitid: str,
+        commit_yaml: dict,
+        processing_results: list[ProcessingResult],
+        report_code: str | None = None,
+        empty_upload=None,
+        **kwargs,
+    ):
+        lock_manager = LockManager(
+            repoid=repoid,
+            commitid=commitid,
+            # Currently hardcoded to coverage, should be variable to the report type
+            report_type=ReportType.COVERAGE,
+            # TODO: Ask if we need a specific timeout hard timeout for this task, unsure
+            lock_timeout=max(80, self.hard_time_limit_task),
+        )
+        try:
+            lock_acquired = False
+            with lock_manager.locked(
+                lock_type=LockType.NOTIFICATION_ORCHESTRATOR,
+                retry_num=self.request.retries,
+            ):
+                lock_acquired = True
+                return self.run_impl_within_lock(
+                    db_session,
+                    repoid=repoid,
+                    commitid=commitid,
+                    commit_yaml=commit_yaml,
+                    processing_results=processing_results,
+                    report_code=report_code,
+                    empty_upload=empty_upload,
+                    **kwargs,
+                )
+        # TODO: What should happen when there's a lock error? I think this should change to some retrying mechanism
+        # Should it be a LockRetry or a LockError?
+        except LockRetry as err:
+            (
+                log.info(
+                    "Not notifying because there is another notification already happening",
+                    extra=dict(
+                        repoid=repoid,
+                        commitid=commitid,
+                        error_type=type(err),
+                        lock_acquired=lock_acquired,
+                    ),
+                ),
+            )
+            UploadFlow.log(UploadFlow.NOTIF_LOCK_ERROR)
+            return {
+                "notified": False,
+                "notifications": None,
+                "reason": "unobtainable_lock",
+            }
+
+    def run_impl_within_lock(
+        self,
+        db_session: Session,
+        *args,
+        repoid: int,
+        commitid: str,
+        commit_yaml: dict,
+        processing_results: list[ProcessingResult],
+        report_code: str | None = None,
+        empty_upload=None,
+        **kwargs,
+    ):
+        repoid = int(repoid)
+        commit = (
+            db_session.query(Commit)
+            .filter(Commit.repoid == repoid, Commit.commitid == commitid)
+            .first()
+        )
+        assert commit, "Commit not found in database."
+        commit_yaml = UserYaml(commit_yaml)
+
+        log_extra_dict = {
```

Comment:
I pass around this log bw functions, but there might be a better strategy to do so, perhaps leveraging the context 👀